src/jit/importer.cpp

```diff
@@ -3344,16 +3344,21 @@ GenTreePtr Compiler::impIntrinsic(CORINFO_CLASS_HANDLE  clsHnd,
 
             op1 = nullptr;
 
-#ifdef LEGACY_BACKEND
+#if defined(LEGACY_BACKEND)
             if (IsTargetIntrinsic(intrinsicID))
-#else
+#elif !defined(_TARGET_X86_)
             // Intrinsics that are not implemented directly by target instructions will
             // be re-materialized as users calls in rationalizer. For prefixed tail calls,
             // don't do this optimization, because
             //  a) For back compatibility reasons on desktop.Net 4.6 / 4.6.1
             //  b) It will be non-trivial task or too late to re-materialize a surviving
             //     tail prefixed GT_INTRINSIC as tail call in rationalizer.
             if (!IsIntrinsicImplementedByUserCall(intrinsicID) || !tailCall)
+#else
+            // On x86 RyuJIT, importing intrinsics that are implemented as user calls can cause incorrect calculation
+            // of the depth of the stack if these intrinsics are used as arguments to another call. This causes bad
+            // code generation for certain EH constructs.
+            if (!IsIntrinsicImplementedByUserCall(intrinsicID))
```

Comment:
I think that we should be able to deal with this in the rationalizer at the time it changes them back into user calls. As you say, it is probably not the most straightforward fix, but is probably a better way to handle it. It's a bit unfortunate to be dealing with stack depth so early in the JIT.

---
In reply to: [89024918](https://github.com/dotnet/coreclr/pull/8235#discussion_r89024918) [](ancestors = 89024918)